logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -312,30 +317,94 @@ private boolean isOutput(final Vertex vertex) {
+        }
+
+        @SuppressWarnings("rawtypes")
+        public void compute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            unorderedCompute(batch, flush, shutdown);
         }
 
-        Dataset toDataset() {
-            return compiled;
+        @SuppressWarnings("rawtypes")
+        public void unorderedCompute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            @SuppressWarnings("unchecked") final RubyArray<JrubyEventExtLibrary.RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+            copyNonCancelledEvents(compiledFilters.compute(batch, flush, shutdown), outputBatch);
+            compiledFilters.clear();
+            compiledOutputs.compute(outputBatch, flush, shutdown);
+        }
+
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        public void orderedCompute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            final RubyArray<JrubyEventExtLibrary.RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+            // send batch one-by-one as single-element batches down the filters
+            final RubyArray<JrubyEventExtLibrary.RubyEvent> filterBatch = RubyUtil.RUBY.newArray(1);
+            batch.forEach((e) -> {
+                filterBatch.set(0, e);
+                copyNonCancelledEvents(compiledFilters.compute(filterBatch, flush, shutdown), outputBatch);
+                compiledFilters.clear();
+            });
+            compiledOutputs.compute(outputBatch, flush, shutdown);
+        }
+
+        /**
+         * Instantiates the graph of compiled filter section {@link Dataset}.
+         * @return Compiled {@link Dataset} representing the filter section of the pipeline.
+         */
+        private Dataset compileFilters() {
+            final Optional<Vertex> filterOutputSeparator = pipelineIR.getGraph()
+                .vertices()
+                .filter(v -> v instanceof SeparatorVertex)
+                .findFirst();
+            if (!filterOutputSeparator.isPresent()) {
+                // should never get here
+                throw new IllegalStateException("Missing Filter End Vertex");
+            }
+
+            // filter out the QueueVertex if there are no filters
+            final Collection<Vertex> filterNodes = filterOutputSeparator.get().getIncomingVertices()
+                .stream()
+                .filter(v -> v instanceof PluginVertex || v instanceof IfVertex)
+                .collect(Collectors.toList());
+            if (filterNodes.isEmpty()) {
+                return Dataset.IDENTITY;
+            } else {
+                return DatasetCompiler.terminalFilterDataset(filterNodes.stream()
+                    .map(leaf -> {
+                        if (leaf instanceof PluginVertex) {
+                            return filterDataset(leaf, flatten(Collections.emptyList(), leaf));
+                        } else {
+                            // can only be a PluginVertex or an IfVertex
```

Comment:
Can be only an `IfVertex` because the case of `PluginVertex` is checked by if

Reply:
I rephrased fro clarity.